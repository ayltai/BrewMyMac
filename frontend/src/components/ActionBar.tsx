import mixpanel from 'mixpanel-browser';
import { GitHub, } from '@mui/icons-material';
import { AppBar, Box, IconButton, Grid, Link, styled, Toolbar, } from '@mui/material';
import React from 'react';

const ClickableImage = styled('img')`
    cursor : pointer;
`;

const PaddedIconButton = styled(IconButton)(({ theme, }) => ({
    marginLeft : theme.spacing(1),
}));

export const ActionBar = ({
    title,
    centerComponent,
    rightComponent,
} : {
    title?           : string,
    centerComponent? : React.ReactNode,
    rightComponent?  : React.ReactNode,
}) => {
    const handleHomepageClick = () => window.open('/', '_self');

    const handleGitHubClick = () => {
        mixpanel.track('Referral - GitHub');

        window.open('https://github.com/ayltai/BrewMyMac', '_blank');
    };

    return (
        <AppBar
            position='sticky'
            color='inherit'>
            <Toolbar>
                <Grid
                    container
                    width='100%'>
                    <Grid
                        item
                        display='flex'
                        flexDirection='row'
                        xs={3}
                        alignItems='center'>
                        <ClickableImage
                            width={48}
                            height={48}
                            src={require('../LogoSmall.png')}
                            onClick={handleHomepageClick} />
                        <Link
                            marginLeft={2}
                            color='inherit'
                            underline='none'
                            href='/'
                            fontFamily='Inconsolata'
                            fontWeight={600}
                            variant='h4'>
                            {title}
                        </Link>
                        <PaddedIconButton onClick={handleGitHubClick}>
                            <GitHub />
                        </PaddedIconButton>
                    </Grid>
                    <Grid
                        item
                        xs={6}
                        alignItems='center'
                        justifyContent='center'>
                        <Box
                            display='flex'
                            flexDirection='row'>
                            <Box flexGrow={1} />
                            {centerComponent}
                            <Box flexGrow={1} />
                        </Box>
                    </Grid>
                    <Grid
                        item
                        xs={3}
                        alignItems='center'
                        justifyContent='flex-end'>
                        {rightComponent}
                    </Grid>
                </Grid>
            </Toolbar>
        </AppBar>
    );
};
