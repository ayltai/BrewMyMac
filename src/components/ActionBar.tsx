import { GitHub, InfoOutlined, LinkedIn, } from '@mui/icons-material';
import { AppBar, Box, IconButton, Grid, Link, styled, Toolbar, } from '@mui/material';
import { useGA4React, } from 'ga-4-react';
import React, { useState, } from 'react';
import { useTranslation, } from 'react-i18next';

import { About, } from '../screens';

const PaddedIconButton = styled(IconButton)(({ theme, }) => ({
    marginLeft : theme.spacing(1),
}));

export const ActionBar = ({
    title,
    centerComponent,
    rightComponent,
} : {
    title?           : string,
    centerComponent? : React.ReactNode,
    rightComponent?  : React.ReactNode,
}) => {
    const ga = useGA4React();

    const { t, } = useTranslation();

    const [ open, setOpen, ] = useState(false);

    const handleGitHubClick = () => {
        if (ga) ga.event('Click', 'GitHub', 'Referral');

        window.open('https://github.com/ayltai/BrewMyMac', '_blank');
    };

    const handleLinkedInClick = () => {
        if (ga) ga.event('Click', 'LinkedIn', 'Referral');

        window.open(t('url_linkedin'), '_blank');
    };

    const handleAboutClick = () => setOpen(true);

    const handleClose = () => setOpen(false);

    return (
        <>
            <AppBar position='sticky'>
                <Toolbar>
                    <Grid
                        container
                        width='100%'>
                        <Grid
                            item
                            display='flex'
                            flexDirection='row'
                            xs={3}
                            alignItems='center'>
                            <Link
                                color='inherit'
                                underline='none'
                                href='/'
                                variant='h5'>
                                {title}
                            </Link>
                            <PaddedIconButton onClick={handleGitHubClick}>
                                <GitHub />
                            </PaddedIconButton>
                            <PaddedIconButton onClick={handleLinkedInClick}>
                                <LinkedIn />
                            </PaddedIconButton>
                            <PaddedIconButton onClick={handleAboutClick}>
                                <InfoOutlined />
                            </PaddedIconButton>
                        </Grid>
                        <Grid
                            item
                            xs={6}
                            justifyContent='center'>
                            <Box
                                display='flex'
                                flexDirection='row'>
                                <Box flexGrow={1} />
                                {centerComponent}
                                <Box flexGrow={1} />
                            </Box>
                        </Grid>
                        <Grid
                            item
                            xs={3}
                            justifyContent='flex-end'>
                            {rightComponent}
                        </Grid>
                    </Grid>
                </Toolbar>
            </AppBar>
            {open && <About onClose={handleClose} />}
        </>
    );
};
