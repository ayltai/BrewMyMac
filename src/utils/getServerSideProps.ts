import { serverSideTranslations, } from 'next-i18next/serverSideTranslations';
import nextI18NextConfig from '../../next-i18next.config';

export default async function getServerSideProps({
    locale,
} : {
    locale : string,
}) {
    return {
        props : {
            ...(await serverSideTranslations(locale, [
                'common',
            ], nextI18NextConfig)),
        },
    };
}
