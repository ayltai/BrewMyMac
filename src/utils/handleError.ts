import * as Sentry from '@sentry/nextjs';

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};
