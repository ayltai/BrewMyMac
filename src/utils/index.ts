export { createEmotionCache, } from './createEmotionCache';
export { handleError, } from './handleError';
