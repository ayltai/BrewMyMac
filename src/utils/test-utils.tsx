import { createTheme, ThemeProvider, } from '@mui/material/styles';
import { configureStore, PreloadedState, Store, } from '@reduxjs/toolkit';
import '@testing-library/jest-dom/extend-expect';
import { act, render, } from '@testing-library/react';
import React, { ReactNode, } from 'react';
import { Provider, } from 'react-redux';

import { homebrewApi, } from '../apis';
import { preferencesReducer, shoppingCartReducer, store as reduxStore, } from '../states';
import { palette, } from '../styles';

import type { HomebrewFormula, } from '../types';

const INITIAL_FILTER = 'firefox';

export const mockApis = async () => {
    const HOMEBREW_API_RESPONSE : HomebrewFormula[] = [
        {
            token     : `${INITIAL_FILTER} homebrew token`,
            name      : [
                `${INITIAL_FILTER} homebrew name`,
            ],
            full_name : `${INITIAL_FILTER} homebrew full name`,
            bottle    : {},
        },
    ];

    const HOMEBREW_CAST_API_RESPONSE : HomebrewFormula[] = [
        {
            token     : `${INITIAL_FILTER} homebrew cask token`,
            name      : [
                `${INITIAL_FILTER} homebrew cask name`,
            ],
            full_name : `${INITIAL_FILTER} homebrew cask full name`,
            bottle    : {},
        },
    ];

    fetchMock.doMock(request => {
        if (request.url.startsWith('https://formulae.brew.sh/api/formula.json')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(HOMEBREW_API_RESPONSE),
            });
        }

        if (request.url.startsWith('https://formulae.brew.sh/api/cask.json')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(HOMEBREW_CAST_API_RESPONSE),
            });
        }

        return Promise.resolve({
            status : 404,
        });
    });

    await act(async () => {
        await reduxStore.dispatch(homebrewApi.endpoints.formula.initiate());
        await reduxStore.dispatch(homebrewApi.endpoints.cask.initiate());
    });
};

const createStore = (preloadedState? : Record<string, any>) => configureStore({
    reducer : {
        preferences  : preferencesReducer,
        shoppingCart : shoppingCartReducer,
    },
    preloadedState,
});

const defaultStore = createStore();

type RootState = ReturnType<typeof defaultStore.getState>;

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : PreloadedState<RootState>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => {
        const theme = createTheme({
            palette : {
                ...palette,
                mode : 'light',
            },
        });

        return (
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </Provider>
        );
    },
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
