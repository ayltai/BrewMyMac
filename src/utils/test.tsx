import { createTheme, ThemeProvider, } from '@mui/material';
import { configureStore, PreloadedState, Store, } from '@reduxjs/toolkit';
import '@testing-library/jest-dom/extend-expect';
import { act, render, } from '@testing-library/react';
import React, { ReactNode, } from 'react';
import { Provider, } from 'react-redux';

import { appStoreApi, homebrewApi, tweakApi, } from '../apis';
import type { AppStoreSearchResults, HomebrewFormula, Tweak, } from '../models';
import { sessionReducer, store as reduxStore, } from '../redux';
import { palette, } from '../styles';

const INITIAL_FILTER = 'firefox';

export const mockApis = async () => {
    const HOMEBREW_API_RESPONSE : HomebrewFormula[] = [
        {
            token     : `${INITIAL_FILTER} homebrew token`,
            name      : [
                `${INITIAL_FILTER} homebrew name`,
            ],
            full_name : `${INITIAL_FILTER} homebrew full name`,
        },
    ];

    const HOMEBREW_CAST_API_RESPONSE : HomebrewFormula[] = [
        {
            token     : `${INITIAL_FILTER} homebrew cask token`,
            name      : [
                `${INITIAL_FILTER} homebrew cask name`,
            ],
            full_name : `${INITIAL_FILTER} homebrew cask full name`,
        },
    ];

    const TWEAKS_API_RESPONSE : Tweak[] = [
        {
            id   : '2',
            name : `${INITIAL_FILTER} tweak name`,
        },
    ];

    const APP_STORE_API_RESPONSE : AppStoreSearchResults = {
        resultCount : 1,
        results     : [
            {
                trackId   : '1',
                trackName : `${INITIAL_FILTER} app store name`,
            },
        ],
    };

    fetchMock.doMock(request => {
        if (request.url.startsWith('https://formulae.brew.sh/api/formula.json')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(HOMEBREW_API_RESPONSE),
            });
        }

        if (request.url.startsWith('https://formulae.brew.sh/api/cask.json')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(HOMEBREW_CAST_API_RESPONSE),
            });
        }

        if (request.url.startsWith('https://raw.githubusercontent.com')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(TWEAKS_API_RESPONSE),
            });
        }

        if (request.url.startsWith('https://search-itunes.vercel.app')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(APP_STORE_API_RESPONSE),
            });
        }

        return Promise.resolve({
            status : 404,
        });
    });

    await act(async () => {
        await reduxStore.dispatch(homebrewApi.endpoints.formula.initiate());
        await reduxStore.dispatch(homebrewApi.endpoints.cask.initiate());
        await reduxStore.dispatch(tweakApi.endpoints.getTweaks.initiate());
        await reduxStore.dispatch(appStoreApi.endpoints.search.initiate(`${INITIAL_FILTER} app store name`));
    });
};

const createStore = (preloadedState? : Record<string, any>) => configureStore({
    reducer : {
        session : sessionReducer,
    },
    preloadedState,
});

const defaultStore = createStore();

type RootState = ReturnType<typeof defaultStore.getState>;

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : PreloadedState<RootState>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => {
        const theme = createTheme({
            palette : {
                ...palette,
                mode : 'light',
            },
        });

        return (
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </Provider>
        );
    },
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
