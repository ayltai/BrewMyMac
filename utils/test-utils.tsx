import { createTheme, ThemeProvider, } from '@mui/material/styles';
import { configureStore, PreloadedState, Store, } from '@reduxjs/toolkit';
import '@testing-library/jest-dom/extend-expect';
import { render, } from '@testing-library/react';
import React, { ReactNode, } from 'react';
import { Provider, } from 'react-redux';

import { preferencesReducer, } from '../states';
import { palette, } from '../styles';

const createStore = (preloadedState? : Record<string, any>) => configureStore({
    reducer : {
        preferences : preferencesReducer,
    },
    preloadedState,
});

const defaultStore = createStore();

type RootState = ReturnType<typeof defaultStore.getState>;

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : PreloadedState<RootState>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => {
        const theme = createTheme({
            palette : {
                ...palette,
                mode : 'light',
            },
        });

        return (
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </Provider>
        );
    },
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
