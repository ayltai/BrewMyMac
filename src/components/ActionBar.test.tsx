import React from 'react';

import { render, } from '../utils/test';

import { ActionBar, } from './ActionBar';

describe('<ActionBar />', () => {
    it('renders correctly', () => expect(render(<ActionBar />).asFragment()).toMatchSnapshot());
});
