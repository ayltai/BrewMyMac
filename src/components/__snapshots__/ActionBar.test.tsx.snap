// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionBar /> renders correctly 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionSticky css-qv5qw8-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container css-t78dm4-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1l0k3fu-MuiGrid-root"
        >
          <a
            class="MuiTypography-root MuiTypography-h5 MuiLink-root MuiLink-underlineNone css-1gcsvp5-MuiTypography-root-MuiLink-root"
            href="/"
          />
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-11b5pux-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="GitHubIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 1.27a11 11 0 00-3.48 21.46c.55.09.73-.28.73-.55v-1.84c-3.03.64-3.67-1.46-3.67-1.46-.55-1.29-1.28-1.65-1.28-1.65-.92-.65.1-.65.1-.65 1.1 0 1.73 1.1 1.73 1.1.92 1.65 2.57 1.2 3.21.92a2 2 0 01.64-1.47c-2.47-.27-5.04-1.19-5.04-5.5 0-1.1.46-2.1 1.2-2.84a3.76 3.76 0 010-2.93s.91-.28 3.11 1.1c1.8-.49 3.7-.49 5.5 0 2.1-1.38 3.02-1.1 3.02-1.1a3.76 3.76 0 010 2.93c.83.74 1.2 1.74 1.2 2.94 0 4.21-2.57 5.13-5.04 5.4.45.37.82.92.82 2.02v3.03c0 .27.1.64.73.55A11 11 0 0012 1.27"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-11b5pux-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="LinkedInIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h14m-.5 15.5v-5.3a3.26 3.26 0 0 0-3.26-3.26c-.85 0-1.84.52-2.32 1.3v-1.11h-2.79v8.37h2.79v-4.93c0-.77.62-1.4 1.39-1.4a1.4 1.4 0 0 1 1.4 1.4v4.93h2.79M6.88 8.56a1.68 1.68 0 0 0 1.68-1.68c0-.93-.75-1.69-1.68-1.69a1.69 1.69 0 0 0-1.69 1.69c0 .93.76 1.68 1.69 1.68m1.39 9.94v-8.37H5.5v8.37h2.77z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-11b5pux-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="InfoOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11 7h2v2h-2zm0 4h2v6h-2zm1-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1dsnw5r-MuiGrid-root"
        >
          <div
            class="MuiBox-root css-1xhj18k"
          >
            <div
              class="MuiBox-root css-i9gxme"
            />
            <div
              class="MuiBox-root css-i9gxme"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-kivoit-MuiGrid-root"
        />
      </div>
    </div>
  </header>
</DocumentFragment>
`;
